go/pkg/casng/streaming_upload.go

```diff
@@ -478,8 +479,133 @@ func (u *uploader) callBatchUpload(ctx context.Context, bundle uploadRequestBund
+			digestReqs[req.Digest] = reqs
+			tags := digestTags[req.Digest]
+			tags = append(tags, req.tag)
+			digestTags[req.Digest] = tags
+			if len(tags) > 1 {
+				// Already in-flight. Release duplicate resources if it's a large file.
+				log.V(3).Infof("[casng] upload.streamer.unified; digest=%s, req=%s, tag=%s, bundle=%d", req.Digest, req.id, req.tag, len(tags))
+				if isLargeFile {
+					u.ioThrottler.release()
+					u.ioLargeThrottler.release()
+				}
+				continue
+			}
+
+			var name string
+			if req.Digest.Size >= u.ioCfg.CompressionSizeThreshold {
+				log.V(3).Infof("[casng] upload.streamer.compress; digest=%s, req=%s, tag=%s", req.Digest, req.id, req.tag)
+				name = MakeCompressedWriteResourceName(u.instanceName, req.Digest.Hash, req.Digest.Size)
+			} else {
+				name = MakeWriteResourceName(u.instanceName, req.Digest.Hash, req.Digest.Size)
+			}
+
+			pending += 1
+			// Block the streamer if the gRPC call is being throttled.
+			startTime := time.Now()
+			if !u.streamThrottle.acquire(u.ctx) {
+				if isLargeFile {
+					u.ioThrottler.release()
+					u.ioLargeThrottler.release()
+				}
+				// Ensure the response is dispatched before aborting.
+				u.workerWg.Add(1)
+				go func(req UploadRequest) {
+					defer u.workerWg.Done()
+					streamResCh <- UploadResponse{Digest: req.Digest, Stats: Stats{BytesRequested: req.Digest.Size}, Err: context.Canceled}
+				}(req)
+				continue
+			}
+			log.V(3).Infof("[casng] upload.streamer.throttle.duration; start=%d, end=%d, tag=%s", startTime.UnixNano(), time.Now().UnixNano(), req.tag)
+			u.workerWg.Add(1)
+			go func(req UploadRequest) {
+				defer u.workerWg.Done()
+				s, err := u.callStream(req.ctx, name, req)
+				// Release before sending on the channel to avoid blocking without actually using the gRPC resources.
+				u.streamThrottle.release()
+				streamResCh <- UploadResponse{Digest: req.Digest, Stats: s, Err: err}
+			}(req)
+		case r := <-streamResCh:
+			startTime := time.Now()
+			r.tags = digestTags[r.Digest]
+			delete(digestTags, r.Digest)
+			r.reqs = digestReqs[r.Digest]
+			delete(digestReqs, r.Digest)
+			u.dispatcherResCh <- r
+			pending -= 1
+			if log.V(3) {
+				log.Infof("[casng] upload.streamer.res; digest=%s, req=%s, tag=%s, pending=%d", r.Digest, strings.Join(r.reqs, "|"), strings.Join(r.tags, "|"), pending)
+			}
+			// Covers waiting on the dispatcher.
+			log.V(3).Infof("[casng] upload.streamer.pub.duration; start=%d, end=%d", startTime.UnixNano(), time.Now().UnixNano())
+		}
+	}
 }
 
 func (u *uploader) callStream(ctx context.Context, name string, req UploadRequest) (stats Stats, err error) {
-	panic("not yet implemented")
+	var reader io.Reader
+
+	// In the off chance that the blob is mis-constructed (more than one content field is set), start
+	// with b.reader to ensure any held locks are released.
+	switch {
+	// Large file.
+	case req.reader != nil:
+		reader = req.reader
+		defer func() {
+			if errClose := req.reader.Close(); err != nil {
```

Comment:
shouldn't be `errClose != nil`?

Reply:
Good catch! `errClose` must be a recent change where `err` was reused before.